### PR TITLE
Add nulls comparison to OrderFilter

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | TODO

Related: #238

TODO:

- [x] Tests
- [ ] Docs

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php [hunk 1]
```diff
@@ -98,6 +114,15 @@ protected function filterProperty(string $property, $direction, QueryBuilder $qu
             list($alias, $field) = $this->addJoinsForNestedProperty($property, $alias, $queryBuilder, $queryNameGenerator);
         }
 
+        if (null !== $nullsComparison = $this->properties[$property]['nulls_comparison'] ?? null) {
+            $nullsDirection = (self::NULLS_SMALLEST === $nullsComparison) ? $direction : ('ASC' === $direction ? 'DESC' : 'ASC');
+
+            $nullRankHiddenField = sprintf('_%s_%s_null_rank', $alias, $field);
+
+            $queryBuilder->addSelect(sprintf('CASE WHEN %s.%s IS NULL THEN 0 ELSE 1 END AS HIDDEN %s', $alias, $field, $nullRankHiddenField));
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: is HIDDEN an alias?

Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: HIDDEN is a DQL keyword that hides the selected field from the result set.

http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: Nice one, thanks. TIL.

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php [hunk 2]
```diff
@@ -98,6 +114,15 @@ protected function filterProperty(string $property, $direction, QueryBuilder $qu
             list($alias, $field) = $this->addJoinsForNestedProperty($property, $alias, $queryBuilder, $queryNameGenerator);
         }
 
+        if (null !== $nullsComparison = $this->properties[$property]['nulls_comparison'] ?? null) {
+            $nullsDirection = (self::NULLS_SMALLEST === $nullsComparison) ? $direction : ('ASC' === $direction ? 'DESC' : 'ASC');
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: are the first parenthesis needed?

Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: You're right. It's not needed.

http://php.net/manual/en/language.operators.precedence.php

Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]:
```php
const NULLS_DIRECTION_MAP = [
    self::NULLS_SMALLEST => [
        'ASC' => 'ASC',
        'DESC' => 'DESC',
    ],
    self::NULLS_LARGEST => [
        'ASC' => 'DESC',
        'DESC' => 'ASC',
    ],
];

...

    $nullsDirection = self::NULLS_DIRECTION_MAP[$nullsComparison][$direction];
```

**Comment by teohhanhui** [comment]
Tests added. :smile:

**Comment by teohhanhui** [comment]
@dunglas Do you think it's worth it to support the shorthand syntax for `nulls_comparison` too?

**Comment by soyuka** [comment]
What do you mean by shortand syntax?

**Comment by teohhanhui** [comment]
> What do you mean by shortand syntax?

In the filter config...

The current shorthand syntax (for `default_direction`):
```yaml
price: 'ASC'
```

Do we want to add this shorthand? (for `nulls_comparison`):
```yaml
price: 'nulls_smallest'
```

Or should we deprecate the shorthand altogether?

**Comment by soyuka** [comment]
Hmm, so `price: 'ASC'` doesn't imply `nulls_smallest`? IMO it would be easier to keep simple and short things, as long as it's documented correctly I see no harm :p.

**Comment by teohhanhui** [comment]
No, we cannot make such assumptions. It'd be wrong in some cases.

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php [hunk 3]
```diff
@@ -57,7 +83,7 @@ public function getDescription(string $resourceClass): array
             $properties = array_fill_keys($this->getClassMetadata($resourceClass)->getFieldNames(), null);
         }
 
-        foreach ($properties as $property => $defaultDirection) {
+        foreach ($properties as $property => $propertyOptions) {
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: why this change ?

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php [hunk 4]
```diff
@@ -81,9 +107,9 @@ protected function filterProperty(string $property, $direction, QueryBuilder $qu
             return;
         }
 
-        if (empty($direction) && isset($this->properties[$property])) {
+        if (empty($direction) && null !== $defaultDirection = $this->properties[$property]['default_direction'] ?? null) {
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: This is a little hard to read no ?

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php (hunk 3 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: Because `$defaultDirection` is now `$propertyOptions['default_direction']`.

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php (hunk 4 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: I welcome suggestions on how to make this more readable.

**Comment by teohhanhui** [comment]
Rebased. I hope this can be merged soon...

**Comment by teohhanhui** [comment]
@theofidry Does this address #238? :smile:

File: src/Bridge/Doctrine/Orm/Filter/OrderFilter.php (hunk 4 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: Can you explain the simple `=` declaration? I don't get how `null !== $defaultDirection = null` works.

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]: Oh wait just got it!

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/OrderFilter.php]:
```
$defaultDirection = $this->properties[$property]['default_direction'] ?? null;

if (empty($direction) && null !== $defaultDirection) {
// ...
}
```
May be more readable same goes below. IMO we can keep those assignations ;).

**Comment by soyuka** [comment]
LGTM! We just need to document this asap or we will forget the feature ever exist :P.

**Comment by theofidry** [comment]
👍
